V7

frncmx opened the pull request:
This PR changes the interface and the algorithm, so it is a major change.
I try to do small commits with meaningful commit messages, so you might want to go commit by commit.

Summary:
- Change signing algo to `PS512` as recommended by RFC, see commit message for details.
- Fetch and store public key at initialization, so locks are gone and the lib provides feedback on init, if AWS config is wrong.
- Replace the unmaintainable Goblin tests.
- Bump dependencies.
- Update Godoc.

The improvements recommended from #6 will be done in an upcoming PR.

frncmx committed:
Bump MAJOR version to v7

The signing algorithm will be changed.

frncmx committed:
test: drop Goblin tests

Goblin tests appeal to the same beauty standard as Goblins from fantasy.
Navigation is broken.
There were globals everywhere with execution hooks making the flow impossible to follow.
Fixing even a simple failure can cause a headache.

We test to make change easy and possible with confidence.
The suite tested internals making internal changes, refactors hard.
It mocked everything so it actually did not increase confidence.

After spending 10 minutes on trying to fix a single test it was enough.
LocalStack already adds higher than ever confidence.
Let's have a lighter suite and add test - that increase confidence - on demand.
Fox example, error scenarios might worth some coverage.

frncmx committed:
fix: error not visible when KMS signing fails

That can be useful for debugging.

frncmx committed:
Retrieve public key from KMS on initialization

That avoids locks and tells us on init if something is wrong with the client.

frncmx committed:
doc: update Godoc comments

frncmx committed:
Change signing algorithm from RS512 to PS512

https://datatracker.ietf.org/doc/html/rfc8017#section-8

As per recommended by the RFC, PS512 is required to be used for new applications.
RS512 is still considered to be secure for signing, not for encryption.
But PS512 is considered more secure.

frncmx committed:
chore(deps): bump all dependencies to latest

frncmx committed:
chore(deps): swap github.com/pkg/errors to standard library

The repo is a public archive, and standard library has now to needed functionality.

frncmx committed:
Prefix error messages with library name

So it is easier to identify the error source.

frncmx committed:
test: add new suite of tests after dropping Goblin tests

Test that error conditions do not blow up the library.
Since dropping the Goblin suite we did not test error handling.

frncmx committed:
Return value type from kmsjwt.New()

As the type does not require anymore the use of pointers.

frncmx committed:
refactor(test): move test helpers

It feels the test code more belonged to kms_test.go.
This has the benefit of removing some noise from between the tests.
So hopefully the tests are easier to read.

eliecharra commented on go.mod:
```diff
@@ -1,45 +1,31 @@
-module github.com/spacelift-io/kmsjwt/v6
+module github.com/spacelift-io/kmsjwt/v7
```
I don't have a lot of experience with major version bumps. Should we create a v6 branch or subfolder or something like that for the v6 module to continue working?

frncmx replied:
No need. The tooling uses tags. If we need a patch release we just checkout the tag, start a branch from there and add a new tag when done. The branch can be even deleted after.

eliecharra replied:
Good to know, thanks!

eliecharra commented on kmsjwt.go:
```diff
@@ -5,105 +5,97 @@ import (
 
-func (k *KMSJWT) Alg() string {
+// Alg returns the signing algorithm as defined in https://datatracker.ietf.org/doc/html/rfc7518#section-3.1.
+func (k KMSJWT) Alg() string {
```
Any reason not to use pointer receivers?

```suggestion
func (k *KMSJWT) Alg() string {
```

frncmx replied:
I usually approach this form the other direction. Any reason to use pointers?

Since the type does not need to modify the struct and the struct is small I defaulted to value receiver.

> For types such as basic types, slices, and small structs, a value receiver is very cheap so unless the semantics of the method requires a pointer, a value receiver is efficient and clear.

https://go.dev/doc/faq#methods_on_values_or_pointers

eliecharra replied:
I just thought we should avoid a copy of the pubkey. This struct does not look like a "basic" type to me, but that's a bit subjective.

Not a strong opinion, though. 

I'm usually more on the other side of the road, as it's a safer default in terms of memory usage. That's easy to forgot to switch to pointer receiver if we add more heavy fields to it.

frncmx replied:
The `rsa.PublicKey` is actually a pointer, so small in the memory. Maybe it is somewhat strange to keep a pointer in a non-pointer struct. 🤔

I am also thinking now about extendability of the struct without breaking the interface. Because of that maybe it would worth the use of pointers. But I think we should be ok for now.

eliecharra replied:
> The rsa.PublicKey is actually a pointer

Ha indeed that's an interface 👍🏻 So I agree that's not a big deal, we could probably move on!

frncmx replied:
We don't have a user base, I think (counting stars and forks is easy 😆). So I am fine to err on the value type side.

eliecharra commented on kmsjwt.go:
```diff
@@ -5,105 +5,97 @@ import (
 	ctx, ok := key.(context.Context)
 	if !ok {
-		return "", jwt.ErrInvalidKeyType
+		return "", fmt.Errorf("kmsjwt sign: %w", jwt.ErrInvalidKeyType)
```
Are you sure about that move? 

It's maybe easier for the caller to assert on `errors.As(err, jwt.ErrInvalidKeyType)` rather than matching the string in a generic error. 
Maybe you can return `errors.join()` if you want to enrich the error message. That being said, I'm not sure if adding `kmsjwt sign` prefix is really helpful here.

frncmx replied:
`fmt.Errorf` wraps the error when `%w` is used. So `errors.As()`, `errors.Is()` works as you except.

I don't think one can use `errors.As` with a sentinel error as that modifies the target variable. But here is an example with `Is()`: https://go.dev/play/p/jW3rDdUoJaT?v=

So I just added extra context while kept the original error accessible.

eliecharra replied:
Oh I did now knew about that! That's neat 👍🏻

eliecharra commented on kmsjwt.go:
```diff
@@ -5,105 +5,97 @@ import (
 
 	hash := signingMethod.Hash.New()
-	hash.Write([]byte(signingString))
+	_, _ = hash.Write([]byte(signingString))
```
Not sure what could goes wrong here, but maybe we should catch and return a wrapped error here?

frncmx replied:
I had mixed feeling about this. The "ignore" was there before, I just made it explicit.

I checked and found out:
- The same snippet is present in all built-in signers, e.g., [jwt - RSA PSS](https://github.com/golang-jwt/jwt/blob/b05644bf946495734b07010a9af7e357a1f239d8/rsa_pss.go#L127)
- The hasher indeed does not seem to return any error, just it implements `io.Writer`, see: [stdlib SHA512](https://cs.opensource.google/go/go/+/refs/tags/go1.23.5:src/crypto/sha512/sha512.go;l=253)

So I decided it should be OK. I guess a comment would be useful.

eliecharra replied:
Yeah, I thought that since we're bumping to a major, introducing behavioral BC break is fine, and that it can't hurt to report that. 
But as I said, I don't feel super strong about it, since I'm not sure what can go wrong here.
That will probably fail just after anyway, it was just in order to make any potential troubleshooting session a little bit easier (maybe 😅)

frncmx replied:
I return an error now. Even though I think that bath will be never exercised, still looks/feels better.

eliecharra commented on kmsjwt.go:
```diff
@@ -5,105 +5,97 @@ import (
-	publicKey, err := k.getPublicKey(ctx)
 	if err != nil {
-		return err
+		return "", fmt.Errorf("kmsjwt signing with KMS: %w", err)
```
Maybe just in order to preserve the original error

```suggestion
		return "", errors.Join(errors.New("kmsjwt signing with KMS"), err)
```

frncmx replied:
See my previous answer about preserving the error. I prefer `fmt.Errorf("%w")` as that wrap. I don't want to have multiple errors with `errors.Join()`. I would use `errors.Join()` if I provided sentinel errors that the user can assert on with `errors.Is()`.

eliecharra commented on ksmjwt_test.go:
```diff
@@ -0,0 +1,126 @@
+	// Valid values: https://datatracker.ietf.org/doc/html/rfc7518#section-3.1
+	const want = "PS512"
+	signer, _ := newSignerAndStub(t)
+	assert.Equal(t, want, signer.Alg(), "algorithm changed, that's MAJOR change")
```
👍🏻

eliecharra commented on kmsjwt_test.go:
```diff
@@ -1,202 +0,0 @@
-)
-
-func TestKMSJWT(t *testing.T) {
-	g := goblin.Goblin(t)
```
❤️

eliecharra reviewed (approved):
Looks good overall, just left a comment about the module version bump 🙏🏻 

And also a few minor comments regarding errors handling.

frncmx committed:
Return error if hash.Write fails in Sign

This should not happen, as the hash implementations - that I checked - don't return an error.
Yet, the `if` on error makes the code look more trustworthy.
